[3주차 기본/심화/공유 과제] 숫자야구 게임 + 깃허브 검색 페이지

shinjigu opened the pull request:
## 🤙🏻 구현 기능 명세
## 💡 기본 과제

- [x] Context API, 전역상태 라이브러리 사용 X (ThemeProvider 제외)
- [x] CSS라이브러리 사용하기
- [x] 외부 UI 라이브러리 없이 직접 컴포넌트를 구현합니다.

1. **헤더**
   - [x] 헤더에는 제목과 2개의 탭(깃허브 검색, 숫쟈야구 게임)이 위치한다.
   - [x] 탭을 클릭하면 각 탭에 맞는 페이지를 렌더링 한다. (라우팅 X, URL은 안바뀜)

2. **숫자야구 게임**
   - [x] 게임이 시작되면 무작위 3자리 정답 숫자를 생성한다.
   - [x] 정답을 맞춘경우 3초후 게임을 초기화 한다.
      
   2.1 **Input**
     - [x] '최대 3자리'의 '숫자'만 입력 가능하다.
	- [x] 중복된 숫자는 입력이 불가능하다.
	   
   2.2 **Message(사용자의 입력값에 대한 정보를 출력)**
     - [x] '최대 3자리'의 '숫자'가 아니라면, 경고 문구를 출력한다.
	- [x] 정답을 맞춘경우 게임 승리 문구를 출력한다.
	- [x] 오답인 경우 게임 결과를 출력한다 (ex. 1스트라이크 1볼)
	   
   2.3 **List(게임 진행 상황을 출력)**
     - [x] 이전 시도 결과를 출력한다. (ex. 123 - 1스트라이크 1볼)
   - [x] 게임이 초기화되면 리스트도 초기화된다.
	   
3. **깃허브 검색 기능**
   - [x] 깃허브 공공 API를 이용한다. (세부 설명에 제공되는 코드 활용)
   
   3.1 **Input**
   	 - [x] 제출 시 GitHub API를 통해 해당 아이디의 프로필 정보를 요청한다.
   	 
   3.2 **Card**
   	 - [x] 검색 성공시 사용자 정보를 출력한다.
   	 - [x] 출력 항목 -> (아바타 이미지 (avatar_url), 이름 (name), 깃허브 아이디 (login), 자기소개 (bio), 깃허브 주소 (html_url), 팔로워 수 (followers), 팔로잉 수 (following)
   	 - [x] x 버튼을 누르면 사용자 정보가 사라진다.
   	 - [x] 사용자 이름 or 아바타 이미지를 클릭하면 해당 깃허브 Url로 이동한다. (새로운 탭이 생겨야함)
   	 
   3.3 **최근 검색어**
   	 - [x] 검색한 아이디는 최근 검색어로 저장된다.
   	 - [x] 브라우저를 새로고침해도 검색 기록은 유지된다. (localStorage 활용)
   	 - [x] x 버튼을 누르면 해당 최근 검색어는 삭제된다. (localStorage에서도 삭제)
   	
---

## 🔥 심화 과제

1. **숫자야구 게임**
   - [x] 시도 횟수는 최대 10회로 제한한다.
   - [x] 시도 횟수가 10번이 넘어가면 5초후 게임을 초기화 한다.
   - [x] 시도 횟수가 10번이 넘어가면, 게임패배 문구를 출력한다.

2. **깃허브 검색 기능**
	 - [x] 최근 검색어를 클릭하면 해당 아이디로 재검색이 실행된다.
	 - [x] 최근 검색어는 최대 3개까지만 유지되며, 최신 검색어가 가장 오른쪽에 위치한다.
	 - [x] 중복된 검색어는 최근 검색어에 저장되지않는다.

3. **깃허브 검색 userInfo status에 따른 분기처리**
   - [x] idle, pending, resolved, rejected에 대한 분기 처리를 진행한다.
   - [x] rejected인 경우(검색 결과가 없는 경우) 결과를 찾을 수 없다는 메세지를 출력한다.
   - [x] pending인 경우(로딩 중인 경우) 스피너를 출력한다.


## 공유과제

### 제목: Tailwind CSS 4.0 설정 문제 해결하기: PostCSS 플러그인 오류부터 Vite 설정까지

**_링크 첨부_** : https://sweeb.tistory.com/90

<br />

## 🎨 CSS 라이브러리 선택 이유
Tailwind CSS를 선택한 이유는 이전에 한 번 사용해 본 적 있어서 제대로 또 해보고 싶었기 때문입니다 ! 유틸리티 클래스 기반의 접근 방식이 컴포넌트 기반 개발과 잘 맞아 개발 속도를 높이고, 디자인 일관성을 유지하며, 반응형 디자인을 쉽게 구현할 수 있어 이번 프로젝트에 적합하다고 판단했습니다. 또한 별도의 CSS 파일 작성 없이도 빠르게 UI를 구현할 수 있어 선택하게 되었습니다.

<br />

## 🚀 내가 새로 알게 된 점
- Tailwind CSS의 유틸리티 클래스 방식이 컴포넌트 기반 개발에 매우 효율적임을 다시 한번 느꼈습니다 !!
- 커스텀 훅(useBaseballGame, useGitHubSearch)을 통해 비즈니스 로직과 UI를 분리하니 코드 가독성과 유지 보수성이 크게 향상된다는 것을 배웠습니다.
- 게임 로직 설계 시 상태를 enum처럼 활용하면 조건부 렌더링이 명확해진다는 점을 배웠습니다.

<br />

## 🤔 구현 과정에서의 어려웠던/고민했던 부분
- 과제를 급하게 하느라 심화 과제 2,3번을 구현하지 못했습니다. merge 전에 공부해서 GitHub 검색 관련 심화 기능을 완성하고 싶습니다 🥲
  - -> 심화 과제 2,3번 추가했습니다 !
- Tailwind CSS 설정 과정에서 어려움이 있었습니다. 처음에는 정식 설치 방법으로 진행했으나 PostCSS 플러그인 문제가 발생해 `<script src="https://cdn.tailwindcss.com"></script>`를 HTML에 추가하는 방법으로 해결했습니다. 하지만 이 방법은 임시방편 같아서 찾아보고 추후 개선할 예정입니다 !! 
  - -> @tailwindcss/postcss 패키지를 설치한 후, postcss.config.js 파일을 ES 모듈 형식으로 수정하고, CSS 파일의 지시문을 `@import "tailwindcss"` 방식으로 변경했습니다. 또한 Vite 환경에 최적화하기 위해 @tailwindcss/vite 플러그인을 설치하고 vite.config.js에 추가하여 해결했습니다. 이 문제와 해결 과정에 대한 자세한 내용은 3주차 공유 과제에 적어두었습니다 ☺️

<br />

## ⏳ 소요 시간

- `10h`

<br />

## 🤳🏻 구현 결과물
### 1. 헤더
https://github.com/user-attachments/assets/2c74a760-c8a6-4175-a734-d22d879d0f17

### 2. 숫자야구 게임
#### 정답을 맞춘 경우 & 유효성 검사, 심화 - 시도횟수를 초과한 경우
https://github.com/user-attachments/assets/ee79340c-e240-4062-9594-b7695fef690f

### 3. 깃허브 검색
#### 3.1 깃허브 검색 - Card
https://github.com/user-attachments/assets/24864e61-5301-45b2-87ab-ec849b0bf342

#### 3.2 깃허브 검색 - 최근 검색어
https://github.com/user-attachments/assets/63e19d7a-f0ae-4c84-b5ea-4589a9f992f9

#### 3.3 (심화) 깃허브 검색 userInfo status에 따른 분기처리
https://github.com/user-attachments/assets/2b1b36f0-c270-4d54-a078-af0a743e3d5f

<!-- 스크린샷, gif, 동영상, 배포링크 등 자유롭게 첨부해주세요. 구현한 기능이 모두 결과물에 포함되도록 구체적으로 작성해주세요. -->
<!-- 애니메이션이나 동적 UI 확인이 필요한 경우에는 스크린샷은 인정하지 않아요. -->

shinjigu committed:
feat: 프로젝트 초기 설정 및 Tailwind CSS 구성

- Vite와 React 프로젝트 생성
- Tailwind CSS 패키지 설치
- tailwind.config.js 파일 생성
- postcss.config.js 파일 생성
- index.css에 Tailwind 설정

shinjigu committed:
feat: 헤더 컴포넌트 및 앱 레이아웃 구현

- 프로젝트 메인 헤더 디자인 구현
- 탭 전환 기능 추가 (깃허브 검색, 숫자야구 게임)
- 앱 레이아웃 구조 설정

shinjigu committed:
fix: Tailwind CSS PostCSS 플러그인 설정 수정

- @tailwindcss/postcss 패키지 설치
- postcss.config.js 파일 업데이트하여 최신 Tailwind CSS v4 지원
- CSS 스타일 적용 오류 해결

shinjigu committed:
feat: GitHub 검색 기능 구현

- 검색창(SearchBar) 컴포넌트 구현
- 최근 검색어(RecentSearches) 기능 구현 및 localStorage 연동
- 사용자 정보 카드(UserCard) 컴포넌트 구현
- GitHub API 연동을 위한 useGitHubSearch 훅 구현
- 진행 상태에 따른 UI 렌더링 처리

shinjigu committed:
fix: Tailwind CSS 적용 문제 해결을 위한 CDN 추가

- index.html에 Tailwind CSS CDN 추가
- 로컬 Tailwind CSS 설정 문제 우회
- 스타일 적용 정상화

shinjigu committed:
feat: 숫자야구 게임 기본 기능 구현

- 게임 로직 및 커스텀 훅(useBaseballGame) 구현
- 게임 컴포넌트(GameInput, GameMessage, GameHistoryList) 구현
- 랜덤 3자리 정답 생성 및 유효성 검사 기능 추가
- 스트라이크, 볼, 아웃 판정 로직 구현
- 게임 진행 기록 및 결과 표시 추가
- 정답 맞힌 경우 3초 후 게임 초기화 기능 구현

shinjigu committed:
feat: 숫자야구 게임 심화 기능 구현

- 시도 횟수 10회 제한 기능 추가
- 게임 패배 시 메시지 출력 및 5초 후 초기화 기능 구현
- 시도 횟수 표시 UI 추가

shinjigu committed:
fix: Tailwind CSS 4 PostCSS 플러그인 문제 해결

- @tailwindcss/postcss 패키지 설치
- postcss.config.js 파일 ES 모듈 형식으로 업데이트

shinjigu committed:
style: 깃허브 검색 및 숫자야구 게임 SearchBar 스타일 통일

shinjigu committed:
fix: Tailwind CSS 4.0 설정 문제 해결

- PostCSS 플러그인 오류 해결 (@tailwindcss/postcss 패키지 설치)
- postcss.config.js 파일 ES 모듈 형식으로 수정
- index.css @import 구문으로 변경
- Vite 설정에 @tailwindcss/vite 플러그인 추가

Jeong-Ag commented on week3/assignment/src/components/BaseballGame/BaseballGamePage.jsx:
```diff
@@ -0,0 +1,31 @@
+// src/components/BaseballGame/BaseballGamePage.jsx
```
불필요한 주석은 삭제해주시면 좋을 것 같아요!!

Jeong-Ag commented on week3/assignment/src/components/BaseballGame/GameHistoryList.jsx:
```diff
@@ -0,0 +1,21 @@
+    <div className="space-y-2">
+      {guesses.map((guess, index) => (
+        <div 
+          key={index} 
```
배열의 인덱스가 key값으로 사용되고 있는데 데이터가 변경되는 경우 DOM 최적화에 문제가 생길 수 있어요. 객체에 고유한 식별자가 있다면 해당 값으로 key로 사용하는 것이 더 안전해요. 아래 자료 참고하시면 좋을 것 같아요!

ref: https://velog.io/@phjjj/React-map%EC%9C%BC%EB%A1%9C-%EC%BB%B4%ED%8F%AC%EB%84%8C%ED%8A%B8-%ED%98%B8%EC%B6%9C%EC%8B%9C-key%EB%A5%BC-index%EB%A1%9C-%EB%91%90%EB%A9%B4-%EC%95%88%EB%90%98%EB%8A%94-%EC%9D%B4%EC%9C%A0

shinjigu replied:
@Jeong-Ag 좋은 피드백 감사합니다 ☺️ 공유해 주신 자료를 통해 인덱스를 key로 사용하면 항목 추가/삭제 시 DOM 업데이트가 비효율적으로 이루어지고, 특히 input과 같은 상태를 가진 컴포넌트에서 문제가 발생할 수 있다는 점을 배웠습니다 !!
useBaseballGame 훅에서 guess 객체 생성 시 타임스탬프를 활용한 고유 ID를 추가하고, GameHistoryList 컴포넌트에서는 그 ID를 key로 사용하도록 수정하겠습니다 !

Jeong-Ag reviewed (changes requested):
우선 두 개만 먼저 남겼어요 ㅎㅎ ... 다른 코드는 목요일까지 봐야지 ~~~

shinjigu committed:
refactor: 하드코딩된 값을 상수로 분리

- 게임 상태(GAME_STATUS)와 설정(GAME_SETTINGS) 상수화
- API 상태(API_STATUS)와 로컬 스토리지 키(STORAGE_KEYS) 상수화
- 게임 메시지(GAME_MESSAGES) 상수화

shinjigu committed:
refactor: 유틸리티 함수 분리 및 도메인별 구성

- 게임 로직 관련 함수를 baseballUtils.js로 분리
- 로컬 스토리지 관련 함수를 githubUtils.js로 분리
- 훅 파일에서 비즈니스 로직 제거하고 유틸리티 함수 사용

shinjigu committed:
refactor: gameState 객체를 개별 상태로 분리

- 하나의 큰 gameState 객체를 여러 개의 작은 상태로 분리
- answer, guesses, message, status, attempts 개별 상태 변수로 관리

shinjigu committed:
refactor: 폴더 구조 변경

- 공통 컴포넌트를 common 폴더로 분리
- 메인 화면 컴포넌트를 views 폴더로 이동
- views 별 하위 컴포넌트 분리 (baseball, github)

shinjigu committed:
refactor: 공통 UI 컴포넌트 생성

- Button, Input 컴포넌트

shinjigu committed:
refactor: useGitHubSearch 훅의 상태 관리 개선

- userInfo 객체를 status와 userData 개별 상태로 분리

shinjigu committed:
refactor: 코드 정리 및 개선

- GameHistoryList 컴포넌트에서 불필요한 빈 공간 요소 제거하고 Grid 레이아웃 적용
- 불필요한 주석 제거

shinjigu committed:
refactor: InputForm 공통 컴포넌트 추가 및 중복 컴포넌트 제거

- InputForm 공통 컴포넌트 추가
- SearchBar, GameInput 컴포넌트 제거

shinjigu committed:
fix: BaseballGame 컴포넌트에 GAME_STATUS 상수 import문 추가

shinjigu committed:
feat: 접근성 개선을 위한 ARIA 속성 추가

shinjigu committed:
feat: 심화 과제 2 - 깃허브 검색 기능

- 최근 검색어 최대 3개까지만 유지하도록 제한
- 최신 검색어가 가장 오른쪽에 위치하도록 구현
- 중복된 검색어는 저장하지 않도록 처리

shinjigu committed:
feat: 심화 과제 3 - 깃허브 검색 userInfo status에 따른 분기 처리

- API 상태(idle, pending, resolved, rejected)에 따른 분기 처리 추가
- 로딩 상태(pending)일 때 Spinner 컴포넌트 표시
- 오류 상태(rejected)일 때 "결과를 찾을 수 없습니다" 메시지 표시
- StatusContent 컴포넌트로 상태 관련 로직 분리

jogpfls commented on week3/assignment/src/App.jsx:
```diff
@@ -0,0 +1,24 @@
+  const [currentTab, setCurrentTab] = useState('github'); // 기본 탭은 github
+
+  return (
+    <div className="min-h-screen bg-gray-100">
```
Tailwind는 한번도 사용해본적이 없었는데 이렇게 jsx 파일에 바로 사용하는거군요!! Styled-Component만 사용했었는데 이번에 Tailwind CSS에 대해서 배워갑니다 !!

shinjigu replied:
@jogpfls Tailwind CSS는 직접 클래스를 적용해서 스타일링하는 방식이라 처음엔 낯설 수 있지만 사용해 보면 개발 속도가 빨라지는 장점이 있더라고요 ! Styled-Component와는 또 다른 매력이 있는 것 같습니다 !!

jogpfls commented on week3/assignment/src/utils/baseballUtils.js:
```diff
@@ -0,0 +1,56 @@
+  // 숫자만 입력 가능
+  if (!/^\d+$/.test(guess)) {
+    return { valid: false, message: GAME_MESSAGES.NUMERIC_ONLY };
+  }
```
저는 유효성 검사를 안하고 했는데 유효성 검사로 하는게 더 효율적이네요 !! 유효성 검사 잘 사용해봐야겠어요 !

jogpfls commented on week3/assignment/src/App.jsx:
```diff
@@ -0,0 +1,24 @@
+import BaseballGame from './components/views/BaseballGame';
+
+function App() {
+  const [currentTab, setCurrentTab] = useState('github'); // 기본 탭은 github
```
여기에서 currentTab의 github는 따로 상수로 분리하지 않은 이유가 있을까요?!

shinjigu replied:
@jogpfls 좋은 피드백 감사합니다 !! 'github'와 'baseball' 값을 상수로 분리하는 걸 놓쳤네요. 'github'와 'baseball' 값이 App.jsx와 Header.jsx에서 여러 번 사용되고 있어서 TAB_TYPES와 같은 상수 객체로 관리하는 것이 더 좋을 것 같네요. 리팩토링 시 반영하겠습니다 😻

jogpfls commented on week3/assignment/src/components/baseball/GameHistoryList.jsx:
```diff
@@ -0,0 +1,19 @@
+    <div className="space-y-2">
+      {guesses.map((guess, index) => (
+        <div 
+          key={index} 
```
key에 index값을 넣는 것보다 고유한 값을 넣어줘도 좋을 것 같아요 ! 순서가 바뀌지 않아서 문제가 되진 않지만.. 작은 저의 의견입니다...!!

shinjigu replied:
@jogpfls 좋은 피드백 감사합니당 ☺️ UseBaseballGame 훅에서 guess 객체를 생성할 때 타임스탬프 기반 ID를 추가하도록 수정하겠습니다 !

jogpfls commented on week3/assignment/src/hooks/useBaseballGame.js:
```diff
@@ -0,0 +1,95 @@
+import { useState, useEffect, useCallback } from 'react';
```
hooks 파일에 분리하는게 저는 어렵더라고요.. 파일 분리를 너무 잘하셔서 잘배우고 갑니다 !!

jogpfls reviewed (commented):
파일 구조가 잘 분리되어 있어서 술술 읽혔습니다... 그리고 hooks이랑 utils안에 파일이 잘 분리되어 있고 잘 구현되어 있어서 많이 배우고 갑니다 !! 코드도 깔끔해서 계속 읽었어요.. 그리고 tailwind 사용법도 알아갈 수 있어서 좋았습니다 ! 3주차 과제하느라 수고하셨어요 !! 👍

shinjigu committed:
chore: index.html 타이틀 변경

Jeong-Ag commented on week3/assignment/src/components/baseball/GameMessage.jsx:
```diff
@@ -0,0 +1,33 @@
+
+function GameMessage({ message, attempts, maxAttempts, gameStatus }) {
+  // 메시지 스타일 결정
+  const getMessageStyle = () => {
+    if (gameStatus === GAME_STATUS.WON) return 'text-green-600 font-bold';
+    if (gameStatus === GAME_STATUS.LOST) return 'text-red-600 font-bold';
+    return '';
+  };
```
`classNames` 라이브러리를 사용해서 조건부 스타일링을 줘도 좋을 것 같아요~ 
`if` 대신 `&&`를 사용하면 해당되는 조건이 없는 경우에  return을 지정할 필요도 없고 똑같은 기능을 하도록 구현할 수 있을 것 같아서 말씀드려요 :)

ref: https://velog.io/@dooreplay/classNamesCSS-Modules

shinjigu replied:
@Jeong-Ag classNames 라이브러리 추천 감사합니다 ! 공유해 주신 자료를 통해 조건부 스타일링을 더 깔끔하게 처리할 수 있다는 것을 배웠습니다 !!
&& 연산자를 활용하면 가독성도 좋아지고 조건문 없이 더 간결하게 작성할 수 있을 것 같네요. 리팩토링 시 적용해 보겠습니다 👍🏻

Jeong-Ag commented on week3/assignment/src/components/github/UserCard.jsx:
```diff
@@ -0,0 +1,51 @@
+import Button from '../common/Button';
+
+function UserCard({ userInfo, onClose }) {
+  const { data } = userInfo;
```
data를 구조 분해 할당해서 사용하면 요소들을 간략하게 사용할 수 있을 것 같아요!

ref: https://ko.javascript.info/destructuring-assignment

Jeong-Ag commented on week3/assignment/src/constants/baseball.js:
```diff
@@ -0,0 +1,25 @@
+export const GAME_STATUS = {
+  PLAYING: 'playing',
+  WON: 'won',
+  LOST: 'lost'
```
그냥 단순하게 궁금해서 여쭤보는데 과거형을 사용하신 이유가 있으실까요?? 

저도 상수를 사용해서 코드를 작성했는데 게임의 상태이기 때문에 단순 현재형으로 나타냈는데 과거형을 사용하신 이유가 궁금해서 코멘트 남겨봅니당

shinjigu replied:
@Jeong-Ag 게임 상태를 과거형으로 표현한 이유는 사용자 관점에서 '결과'를 나타내기 위함이었습니다 😆 게임에서 이미 이겼다(won) 또는 졌다(lost)라는 결과가 확정된 상태를 표현하고 싶었어요 ㅋㅋㅋㅋㅋ 하지만 현재형으로 표현해도 충분히 의미 전달이 될 것 같습니다 ㅎㅎ

Jeong-Ag commented on week3/assignment/src/components/github/UserCard.jsx:
```diff
@@ -0,0 +1,51 @@
+      </div>
+      <div className="grid grid-cols-2 gap-4">
+        <div className="bg-blue-600 p-4 rounded-lg text-center">
+          <h3 className="font-medium">Followers</h3>
```
`h3` 태그보다는 `p` 태그가 더 적합하다고 생각하는데 h3태그를 사용하신 이유가 있으실까요? 
h는 제목을 나타내는데 사용되는데 해당 값은 제목보다는 단순 텍스트라고 생각되어서 여쭤봐요 🤔

shinjigu replied:
@Jeong-Ag 좋은 피드백 감사합니다 ! 말씀하신 대로 'Followers'는 제목이라기보다 라벨에 가깝네요 !!
의미적으로 p 태그가 더 적합할 것 같습니다. 웹 접근성 측면에서도 제목이 아닌 라벨에는 p 태그를 사용하는 것이 맞겠네요. 수정하겠습니다 !!

Jeong-Ag commented on week3/assignment/src/components/github/Spinner.jsx:
```diff
@@ -0,0 +1,11 @@
+function Spinner() {
+  return (
+    <div className="flex justify-center mt-6">
+      <div className="animate-spin rounded-full h-12 w-12 border-4 border-gray-700 border-t-transparent"></div>
```
self-closing해도 될 것 같아요~

Jeong-Ag commented on week3/assignment/src/components/common/InputForm.jsx:
```diff
@@ -0,0 +1,47 @@
+  };
+
+  const baseClasses = 'w-full px-4 py-3 bg-blue-50 border border-blue-200 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500';
+  const disabledClasses = disabled ? 'disabled:bg-gray-100 disabled:text-gray-500' : '';
```
이 상수도 위의 경우와 똑같다고 생각해요
추가로 disabled인 경우에면 스타일링 적용되도록 작성해주셔서 삼항 연산자 없이 클래스만 넣는건 어떨까요?

Jeong-Ag commented on week3/assignment/src/components/common/InputForm.jsx:
```diff
@@ -0,0 +1,47 @@
+    }
+  };
+
+  const baseClasses = 'w-full px-4 py-3 bg-blue-50 border border-blue-200 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500';
```
해당 스타일링은 기본적으로 사용되니까 상수로 분리 안하고 넣어도 될 것 같아요!

shinjigu replied:
@Jeong-Ag 기본 스타일링은 상수로 분리하지 않고 className 속성에 직접 넣는 것이 더 간결할 것 같네요 수정하겠습니다 !

Jeong-Ag commented on week3/assignment/src/components/common/InputForm.jsx:
```diff
@@ -0,0 +1,47 @@
+    }
+  };
+
+  const baseClasses = 'w-full px-4 py-3 bg-blue-50 border border-blue-200 rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500';
```
해당 스타일링은 기본적으로 사용되니까 상수로 분리 안하고 넣어도 될 것 같아요!

Jeong-Ag commented on week3/assignment/src/components/common/Header.jsx:
```diff
@@ -0,0 +1,28 @@
+        <h1 className="text-xl font-bold mb-4 text-center">⚾ 숫자야구 || 깃허브 검색 🤔</h1>
+        <div className="flex justify-center space-x-4">
+          <Button
+            variant={currentTab === 'github' ? 'secondary' : 'primary'}
```
Button 기본 variant가 primary라서 삼향 연산자 대신에 && 사용하는 건 어떨까용 github면 secondary 전달되게 하면 될 것 같아요

shinjigu replied:
@Jeong-Ag 넵 좋은 제안입니다 !! Button의 기본 variant가 primary니까 삼항 연산자 대신 && 연산자를 사용하면 더 간결해질 것 같습니다 ☺️

Jeong-Ag commented on week3/assignment/src/App.jsx:
```diff
@@ -0,0 +1,24 @@
+    <div className="min-h-screen bg-gray-100">
+      <Header currentTab={currentTab} setCurrentTab={setCurrentTab} />
+      <main className="container mx-auto p-4 max-w-xl">
+        {currentTab === 'github' ? (
+          <GitHubSearch />
+        ) : (
+          <BaseballGame />
+        )}
```
내용이 확장되어 탭이 늘어날 경우를 가정한다면 삼항 연산자보단 &&나 switch 등 여러 컴포넌트를 다루기 좋은 방법으로 리팩토링 해도 좋을 것 같네요 :)

shinjigu replied:
@Jeong-Ag 좋은 제안 감사합니다 ! 말씀하신 대로 탭이 늘어날 경우를 고려하면 삼항 연산자보다는 객체 매핑이나 switch 문을 활용하는 것이 확장성에 더 좋을 것 같습니다. 리팩토링 시 반영하겠습니다 !!

Jeong-Ag reviewed (commented):
컴포넌트에 따라서 잘 분리되어 있어서 완성도가 높다고 생각해요! 폴더 구조도 적합하게 구성되어 있고 커스텀 훅을 사용하신 점도 인상 깊어요 👍 수고 많으셨습니다! 합세 화이띵 ~

shinjigu committed:
refactor: 코드리뷰 피드백 반영

- 탭 타입을 TAB_TYPES 상수로 분리
- 조건부 렌더링을 객체 매핑 방식으로 개선
- 스타일링 코드 간결화 및 조건부 스타일링 개선
- 리스트 아이템에 고유 ID 추가
- 시맨틱 HTML 마크업 개선 (h3 → p 태그)
- 구조 분해 할당을 통한 코드 가독성 향상
- self-closing 태그 적용
- classNames 라이브러리 적용
